lib/src/main/java/graphql/nadel/NextgenEngine.kt

```diff
@@ -388,35 +395,45 @@ internal class NextgenEngine(
         }
 
         return serviceExecResult.copy(
-            data = serviceExecResult.data.let { data ->
-                data.takeIf { transformedQuery.resultKey in data }
-                    ?: mutableMapOf(transformedQuery.resultKey to null)
-            },
+            data = serviceExecResult.data
+                .let { data ->
+                    val newData: MutableJsonMap = mutableMapOf()
+                    topLevelFields.forEach { field ->
+                        newData[field.resultKey] = data[field.resultKey]
+                    }
+                    newData
+                },
         )
     }
 
-    private fun chooseServiceExecution(
+    private fun getServiceExecution(
         service: Service,
-        transformedQuery: ExecutableNormalizedField,
+        topLevelFields: List<ExecutableNormalizedField>,
         hints: NadelExecutionHints,
     ): ServiceExecution {
-        return when {
-            hints.shortCircuitEmptyQuery(service) && onlyTopLevelTypenameField(transformedQuery) ->
-                engineSchemaIntrospectionService.serviceExecution
-            else -> service.serviceExecution
+        if (hints.shortCircuitEmptyQuery(service) && isOnlyTopLevelFieldTypename(topLevelFields)) {
+            return engineSchemaIntrospectionService.serviceExecution
         }
+
+        return service.serviceExecution
     }
 
-    private fun onlyTopLevelTypenameField(executableNormalizedField: ExecutableNormalizedField): Boolean {
-        if (executableNormalizedField.fieldName == TypeNameMetaFieldDef.name) {
+    private fun isOnlyTopLevelFieldTypename(topLevelFields: List<ExecutableNormalizedField>): Boolean {
```

Comment:
This code here was modified a bit.

Context: we wanted to add code that would shortcut query execution IF there was only one field and it was `__typename`

Note: this code is protected by a FF `hints.shortCircuitEmptyQuery(service)` and it's mostly off because it hasn't been fully rolled out.